src/linkml_transformer/transformer/eval_utils.py

```diff
@@ -0,0 +1,199 @@
+"""
+meta-circular interpreter for evaluating python expressions
+
+ - See `<https://stackoverflow.com/questions/2371436/evaluating-a-mathematical-expression-in-a-string>`_
+"""
+
+import ast
+import operator as op
+
+# supported operators
+from typing import Any, List, Tuple
+
+operators = {
+    ast.Add: op.add,
+    ast.Sub: op.sub,
+    ast.Mult: op.mul,
+    ast.Div: op.truediv,
+    ast.Pow: op.pow,
+    ast.BitXor: op.xor,
+    ast.USub: op.neg,
+}
+compare_operators = {ast.Eq: op.eq, ast.Lt: op.lt, ast.LtE: op.le, ast.Gt: op.gt, ast.GtE: op.ge}
+
+
+def eval_conditional(*conds: List[Tuple[bool, Any]]) -> Any:
+    """
+    >>> cond(x < 25 : 'low', x > 25 : 'high', True: 'low')
+    :param subj:
+    :return:
+    """
+    for is_true, val in conds:
+        if is_true:
+            return val
+
+
+funcs = {
+    "max": (True, max),
+    "min": (True, min),
+    "len": (True, len),
+    "str": (False, str),
+    "strlen": (False, len),
+    "case": (False, eval_conditional),
+}
+
+
+class UnsetValueException(Exception):
+    pass
+
+
+def eval_expr(expr: str, **kwargs) -> Any:
+    """
+    Evaluates a given expression, with restricted syntax
+
+    >>> eval_expr('2^6')
+    4
+    >>> eval_expr('2**6')
+    64
+    >>> eval_expr('1 + 2*3**(4^5) / (6 + -7)')
+    -5.0
+
+    Variables:
+
+    variables can be passed
+
+    >>> eval_expr('{x} + {y}', x=1, y=2)
+    3
+
+    Nulls:
+
+    - If a variable is enclosed in {}s then entire expression will eval to None if variable is unset
+
+    >>> eval_expr('{x} + {y}', x=None, y=2)
+    None
+
+    Functions:
+
+    - only a small set of functions are currently supported. All SPARQL functions will be supported in future
+
+    >>> eval_expr('strlen("a" + "bc")')
+    3
+
+    Paths:
+
+    - Expressions such as `person.name` can be used on objects to lookup by attribute/slot
+    - Paths can be chained, e.g. `person.address.street`
+    - Operations on lists are distributed, e.g `container.persons.name` will return a list of names
+    - Similarly `strlen(container.persons.name)` will return a list whose members are the lengths of all names
+
+    :param expr: expression to evaluate
+    """
+    # if kwargs:
+    #    expr = expr.format(**kwargs)
+    if expr == "None":
+        # TODO: do this as part of parsing
+        return None
```

Comment:
changed this from `if "None" in expr` to make it less permissive

Reply:
...although I think it needs to go completely. Not sure if there are cases in the main codebase where this functionality is required.

Reply:
That is the only change to this file.